lib/Sema/CSRanking.cpp

```diff
@@ -446,6 +451,32 @@ static bool isDeclAsSpecializedAs(TypeChecker &tc, DeclContext *dc,
         return inProtocolExtension2;
       }
 
+      // A concrete type member is always more specialised than a protocol
+      // member (bearing in mind that we have already handled the case where
+      // exactly one member is in a protocol extension). Only apply this rule in
+      // Swift 5 mode to better maintain source compatibility under Swift 4
+      // mode.
+      //
+      // Don't apply this rule when comparing two overloads found through
+      // dynamic lookup to ensure we keep cases like this ambiguous:
+      //
+      //    @objc protocol P {
+      //      var i: String { get }
+      //    }
+      //    class C {
+      //      @objc var i: Int { return 0 }
+      //    }
+      //    func foo(_ x: AnyObject) {
+      //      x.i // ensure ambiguous.
+      //    }
+      //
+      if (tc.Context.isSwiftVersionAtLeast(5) && !isDynamicOverloadComparison) {
+        auto *proto1 = dyn_cast<ProtocolDecl>(outerDC1);
+        auto *proto2 = dyn_cast<ProtocolDecl>(outerDC2);
+        if (proto1 != proto2)
```

Comment:
I just ran into an issue with this change.

If you compare two declarations in two different protocols, e.g.:

```
protocol P {
  func f()
}

protocol Q {
  func f()
}
```

it always returns `proto2`.

For the concrete vs. protocol case it doesn't seem (from a quick look at the code) to be doing the right thing either.

Reply:
Ah, I see – not sure how I missed that! We should really be doing an is-a check rather than comparing pointers, i.e:

```
        auto inProto1 = isa<ProtocolDecl>(outerDC1);
        auto inProto2 = isa<ProtocolDecl>(outerDC2);
        if (inProto1 != inProto2)
          return inProto2;
```

I can't immediately think of a test case that trips up the current logic though – for the case you mention of having a protocol refine another, we should be filtering out overridden protocols requirements in name lookup.

Do you have a test case that trips up the current logic? Would be nice to have one for the fix.

> For the concrete vs. protocol case it doesn't seem (from a quick look at the code) to be doing the right thing either.

Could you please elaborate on this? I can't immediately think of a concrete vs. protocol case where it does the wrong thing.

Reply:
I'm experimenting with changes that call `compareDeclarations` on arbitrary overloads in a disjunction to see if they are ordered. I don't have a test case that I know will hit this otherwise.

Reply:
Ah okay, sorry about that! I've submitted a fix: https://github.com/apple/swift/pull/19561